File: node-server/src/services/jsearchSchedulerService.js
import cron from 'node-cron';
import { getDB } from '../database.js';
import { searchAndStoreJsearchJobs } from './jobsService.js';
import { settings } from '../config/settings.js';

const schedulerJobs = new Map();

function cronEnabled() {
  return settings.APP_ENV !== 'production';
}

export async function restoreJsearchSchedulers() {
  if (!cronEnabled()) {
    console.log('[JSearch Scheduler] skipped restore in production (Cloud Run stateless mode)');
    return;
  }

  const db = getDB();
  const active = await db.collection('jsearch_scheduler_meta')
    .find({ enabled: true })
    .toArray();

  for (const state of active) {
    if (!state.config) continue;
    const orgId = state.org_id;
    const config = state.config;
    const jobId = `jsearch_auto_pull:${orgId}`;
    const cronExpr = intervalToCron(config.interval_minutes || 60);
    const task = cron.schedule(cronExpr, () => runJsearchTick(orgId), { timezone: 'UTC' });
    schedulerJobs.set(jobId, task);
    console.log(`[JSearch Scheduler] restored org_id=${orgId} interval=${config.interval_minutes}min`);
  }

  console.log(`[JSearch Scheduler] restored ${active.length} scheduler(s)`);
}

async function saveSchedulerState(orgId, enabled, config = null, lastRunAt = null, lastResult = null) {
  const db = getDB();
  const update = { enabled, updated_at: new Date() };
  if (config) update.config = config;
  if (lastRunAt) update.last_run_at = lastRunAt;
  if (lastResult) update.last_result = lastResult;

  await db.collection('jsearch_scheduler_meta').updateOne(
    { org_id: orgId },
    { $set: update },
    { upsert: true }
  );
}

async function getSchedulerState(orgId) {
  const db = getDB();
  return db.collection('jsearch_scheduler_meta').findOne({ org_id: orgId });
}

async function runJsearchTick(orgId) {
  const state = await getSchedulerState(orgId);
  if (!state?.enabled || !state?.config) return;

  const config = state.config;
  let totalFetched = 0;
  let totalNew = 0;
  let totalStored = 0;
  const errors = [];

  for (const keyword of config.keywords || []) {
    try {
      const result = await searchAndStoreJsearchJobs({
        keywords: keyword,
        location: config.location,
        country: config.country,
        language: config.language,
        numPages: config.num_pages,
        datePosted: config.date_posted,
        workFromHome: config.work_from_home,
        employmentTypes: config.employment_types,
        jobRequirements: config.job_requirements,
        radius: config.radius,
        excludeJobPublishers: config.exclude_job_publishers,
        useCursor: config.use_cursor,
      }, orgId);

      totalFetched += result.fetched_count;
      totalStored += result.stored_count;
      totalNew += result.new_count;
    } catch (err) {
      errors.push({ keyword, error: err.message });
    }
  }

  await saveSchedulerState(orgId, true, null, new Date(), {
    total_fetched: totalFetched,
    total_stored: totalStored,
    total_new: totalNew,
    keywords_run: (config.keywords || []).length,
    errors,
  });
}

function intervalToCron(minutes) {
  if (minutes < 60) return `*/${minutes} * * * *`;
  const hours = Math.floor(minutes / 60);
  return `0 */${hours} * * *`;
}

export async function startJsearchSchedulerForOrg(orgId, config) {
  const jobId = `jsearch_auto_pull:${orgId}`;

  if (!cronEnabled()) {
    await saveSchedulerState(orgId, true, config);
    return {
      enabled: true,
      running: false,
      mode: 'persistent-state-only',
      job_id: jobId,
      interval_minutes: config.interval_minutes,
      next_run_at: null,
      last_run_at: null,
      last_result: null,
      config,
      message: 'Scheduler state saved, but in-memory cron is disabled in production on Cloud Run.',
    };
  }

  if (schedulerJobs.has(jobId)) {
    schedulerJobs.get(jobId).stop();
    schedulerJobs.delete(jobId);
  }

  const cronExpr = intervalToCron(config.interval_minutes);
  const task = cron.schedule(cronExpr, () => runJsearchTick(orgId), { timezone: 'UTC' });
  schedulerJobs.set(jobId, task);

  await saveSchedulerState(orgId, true, config);

  return {
    enabled: true,
    running: true,
    mode: 'in-memory-cron',
    job_id: jobId,
    interval_minutes: config.interval_minutes,
    next_run_at: null,
    last_run_at: null,
    last_result: null,
    config,
  };
}

export async function stopJsearchSchedulerForOrg(orgId) {
  const jobId = `jsearch_auto_pull:${orgId}`;

  if (schedulerJobs.has(jobId)) {
    schedulerJobs.get(jobId).stop();
    schedulerJobs.delete(jobId);
  }

  const state = await getSchedulerState(orgId);
  await saveSchedulerState(orgId, false);

  return {
    enabled: false,
    running: false,
    mode: cronEnabled() ? 'in-memory-cron' : 'persistent-state-only',
    job_id: jobId,
    interval_minutes: state?.config?.interval_minutes || 60,
    next_run_at: null,
    last_run_at: state?.last_run_at || null,
    last_result: state?.last_result || null,
    config: state?.config || null,
  };
}

export async function getJsearchSchedulerStatusForOrg(orgId) {
  const jobId = `jsearch_auto_pull:${orgId}`;
  const state = await getSchedulerState(orgId);

  return {
    enabled: Boolean(state?.enabled),
    running: cronEnabled() ? schedulerJobs.has(jobId) : false,
    mode: cronEnabled() ? 'in-memory-cron' : 'persistent-state-only',
    job_id: jobId,
    interval_minutes: state?.config?.interval_minutes || 60,
    next_run_at: null,
    last_run_at: state?.last_run_at || null,
    last_result: state?.last_result || null,
    config: state?.config || null,
  };
}

File: app/services/jsearch_scheduler_service.py
import logging
from datetime import datetime
from typing import Any, Dict, Optional

from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.interval import IntervalTrigger

from app.database import get_db
from app.models.scraped_job import (
    JSearchRequest,
    JSearchSchedulerConfig,
    JSearchSchedulerStatusResponse,
)
from app.services.jobs_service import search_and_store_jsearch_jobs

logger = logging.getLogger(__name__)

JSEARCH_SCHEDULER_JOB_ID = "jsearch_auto_pull"

# Every keyword is fetched in each of these regions per tick. (location, 2-char country code).
JSEARCH_REGIONS = [("India", "in"), ("United States", "us")]

scheduler = AsyncIOScheduler(timezone="UTC")


async def init_jsearch_scheduler() -> None:
    if not scheduler.running:
        scheduler.start()
        logger.info("JSearch scheduler started")


async def shutdown_jsearch_scheduler() -> None:
    if scheduler.running:
        scheduler.shutdown(wait=False)
        logger.info("JSearch scheduler stopped")


async def _save_scheduler_state(
    org_id: str,
    enabled: bool,
    config: Optional[Dict[str, Any]] = None,
    last_run_at: Optional[datetime] = None,
    last_result: Optional[Dict[str, Any]] = None,
    current_keyword_index: Optional[int] = None,
) -> None:
    db = get_db()
    update_doc: Dict[str, Any] = {"enabled": enabled, "updated_at": datetime.utcnow()}
    if config is not None:
        update_doc["config"] = config
    if last_run_at is not None:
        update_doc["last_run_at"] = last_run_at
    if last_result is not None:
        update_doc["last_result"] = last_result
    if current_keyword_index is not None:
        update_doc["current_keyword_index"] = current_keyword_index

    await db.jsearch_scheduler_meta.update_one(
        {"org_id": org_id},
        {"$set": update_doc},
        upsert=True,
    )


async def _get_scheduler_state(org_id: str) -> Optional[dict]:
    db = get_db()
    return await db.jsearch_scheduler_meta.find_one({"org_id": org_id})


async def _run_jsearch_tick(org_id: str) -> None:
    state = await _get_scheduler_state(org_id)
    if not state or not state.get("enabled") or not state.get("config"):
        logger.info("Scheduler tick skipped org_id=%s — scheduler disabled", org_id)
        return

    config = JSearchSchedulerConfig(**state["config"])
    keywords = [k.strip() for k in config.keywords if k and k.strip()]

    if not keywords:
        logger.info("Scheduler tick skipped org_id=%s — no keywords configured", org_id)
        await _save_scheduler_state(
            org_id=org_id,
            enabled=True,
            last_run_at=datetime.utcnow(),
            last_result={"total_fetched": 0, "keyword_ran": None, "error": "No keywords configured"},
        )
        return

    # One keyword per tick (round-robin), mirroring the Indeed scheduler —
    # firing all keywords every tick multiplied paid JSearch API calls by len(keywords).
    current_index = int(state.get("current_keyword_index", 0)) % len(keywords)
    keyword = keywords[current_index]
    next_index = (current_index + 1) % len(keywords)

    total_fetched = 0
    total_new = 0
    total_stored = 0
    errors = []

    # Fetch every keyword in both India and the USA each tick, regardless of the single
    # location/country stored in the config.
    for location, country in JSEARCH_REGIONS:
        payload = JSearchRequest(
            keywords=keyword,
            location=location,
            country=country,
            language=config.language,
            num_pages=config.num_pages,
            date_posted=config.date_posted,
            work_from_home=config.work_from_home,
            employment_types=config.employment_types,
            job_requirements=config.job_requirements,
            radius=config.radius,
            exclude_job_publishers=config.exclude_job_publishers,
            use_cursor=config.use_cursor,
        )

        logger.info(
            "Scheduler tick keyword=%s org_id=%s location=%s country=%s index=%d next_index=%d",
            keyword, org_id, location, country, current_index, next_index,
        )

        try:
            result = await search_and_store_jsearch_jobs(payload, org_id=org_id)
            total_fetched += result.fetched_count
            total_stored += result.stored_count
            total_new += result.new_count
        except Exception as exc:
            logger.exception(
                "Scheduler tick failed keyword=%s country=%s org_id=%s", keyword, country, org_id
            )
            errors.append({"keyword": keyword, "country": country, "error": str(exc)})

    last_result = {
        "total_fetched": total_fetched,
        "total_stored": total_stored,
        "total_new": total_new,
        "keyword_ran": keyword,
        "keyword_index_ran": current_index,
        "next_keyword_index": next_index,
        "errors": errors,
    }

    logger.info(
        "Scheduler tick complete org_id=%s keyword=%s fetched=%d stored=%d new=%d errors=%d",
        org_id, keyword, total_fetched, total_stored, total_new, len(errors),
    )

    await _save_scheduler_state(
        org_id=org_id,
        enabled=True,
        last_run_at=datetime.utcnow(),
        last_result=last_result,
        current_keyword_index=next_index,
    )

async def start_jsearch_scheduler_for_org(
    org_id: str,
    config: JSearchSchedulerConfig,
) -> JSearchSchedulerStatusResponse:
    await init_jsearch_scheduler()

    job_id = f"{JSEARCH_SCHEDULER_JOB_ID}:{org_id}"
    existing = scheduler.get_job(job_id)
    if existing:
        scheduler.remove_job(job_id)

    scheduler.add_job(
        _run_jsearch_tick,
        trigger=IntervalTrigger(minutes=config.interval_minutes),
        id=job_id,
        kwargs={"org_id": org_id},
        replace_existing=True,
        max_instances=1,
        coalesce=True,
    )

    await _save_scheduler_state(
        org_id=org_id,
        enabled=True,
        config=config.model_dump(),
    )

    job = scheduler.get_job(job_id)

    return JSearchSchedulerStatusResponse(
        enabled=True,
        running=job is not None,
        job_id=job_id,
        interval_minutes=config.interval_minutes,
        next_run_at=job.next_run_time if job else None,
        last_run_at=None,
        last_result=None,
        config=config,
    )


async def stop_jsearch_scheduler_for_org(org_id: str) -> JSearchSchedulerStatusResponse:
    job_id = f"{JSEARCH_SCHEDULER_JOB_ID}:{org_id}"
    job = scheduler.get_job(job_id)
    if job:
        scheduler.remove_job(job_id)

    state = await _get_scheduler_state(org_id)
    await _save_scheduler_state(org_id=org_id, enabled=False)

    config = None
    interval_minutes = 5
    last_run_at = None
    last_result = None

    if state:
        raw_config = state.get("config")
        if raw_config:
            config = JSearchSchedulerConfig(**raw_config)
            interval_minutes = config.interval_minutes
        last_run_at = state.get("last_run_at")
        last_result = state.get("last_result")

    return JSearchSchedulerStatusResponse(
        enabled=False,
        running=False,
        job_id=job_id,
        interval_minutes=interval_minutes,
        next_run_at=None,
        last_run_at=last_run_at,
        last_result=last_result,
        config=config,
    )


async def get_jsearch_scheduler_status_for_org(org_id: str) -> JSearchSchedulerStatusResponse:
    job_id = f"{JSEARCH_SCHEDULER_JOB_ID}:{org_id}"
    job = scheduler.get_job(job_id)
    state = await _get_scheduler_state(org_id)

    config = None
    interval_minutes = 5
    enabled = False
    last_run_at = None
    last_result = None

    if state:
        enabled = bool(state.get("enabled", False))
        if state.get("config"):
            config = JSearchSchedulerConfig(**state["config"])
            interval_minutes = config.interval_minutes
        last_run_at = state.get("last_run_at")
        last_result = state.get("last_result")

    return JSearchSchedulerStatusResponse(
        enabled=enabled,
        running=job is not None,
        job_id=job_id,
        interval_minutes=interval_minutes,
        next_run_at=job.next_run_time if job else None,
        last_run_at=last_run_at,
        last_result=last_result,
        config=config,
    )